File: src/recognition_service/service/dapr_service.py
"""
Module containing code related to a Dapr integration.
"""
import json
from dapr.clients import DaprClient
from service.model.dtos import JsonSerializable


class DaprService:
    """
    A service class for wrapping Dapr functionality.
    """

    @staticmethod
    def publish_event(topic: str, data: JsonSerializable) -> bool:
        """
        Method for publishing an event into the MQ.
        """
        try:
            with DaprClient() as client:
                client.publish_event(
                    pubsub_name="pub-sub",
                    topic_name=topic,
                    data=json.dumps(data.serialize()),
                    data_content_type="application/json"
                )
        except:
            return False
        return True

File: src/webservice/service/dapr_service.go
package service

import (
	"context"
	"recognition-service/service/errors"

	dapr "github.com/dapr/go-sdk/client"
)

// A wrapper service for a Dapr client.
type DaprService struct {
	DaprClient dapr.Client
}

const (
	PUBSUB_NAME string = "pub-sub"
)

// A constructor function for DaprService structure.
func NewDaprService(client dapr.Client) *DaprService {
	return &DaprService{
		DaprClient: client,
	}
}

// Method for publishing an event in the system.
func (srvc DaprService) PublishEvent(ctx context.Context, topic string, data interface{}) error {
	if srvc.DaprClient == nil {
		return errors.ErrDaprIsNotInitialized
	}
	return srvc.DaprClient.PublishEvent(ctx, PUBSUB_NAME, topic, data)
}
